Refactor kubectl methods into utility class, so that we can easily integrate new tests and kube files written in json or other formats.

jayunit100 opened the pull request.

jayunit100 commented:
quick attempt at refactoring e2e kubectl tests, so that they are just importing from a reusable library.
- I've taken the generic functions for testing RC's and moved them to utils.
- created a lambda for examples test which is passed into the utils class, this allows for separation of concerns
- attempted to keep exact test coverage the same, although i didn't confirm this b/c tests don't run perfectly on my current cluster.

jayunit100 committed:
Move core functionality to utils.go, to pave way for incorporating broader testing from examples (push to restart the build)

roberthbailey commented on test/e2e/kubectl.go:
```diff
@@ -17,18 +17,13 @@ limitations under the License.
-	"os/exec"
+	"errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	. "github.com/onsi/ginkgo"
```
The ginkgo import should be last (with a blank line before) as it was before your change. This is consistent with how it's imported in the rest of the e2e test files.

roberthbailey commented on test/e2e/kubectl.go:
```diff
@@ -242,42 +172,9 @@ func getData(c *client.Client, podID string) (*updateDemoData, error) {
 	Logf("got data: %s", body)
 	var data updateDemoData
 	err = json.Unmarshal(body, &data)
```
Check the error after unmarshaling, e.g. if err := json.Unmarshal(...) err != nil { return nil, err }

roberthbailey commented on test/e2e/util.go:
```diff
@@ -17,18 +17,20 @@ limitations under the License.
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/clientauth"
```
Please add a blank line before the '.' imports.

roberthbailey commented on test/e2e/util.go:
```diff
@@ -17,18 +17,20 @@ limitations under the License.
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/clientauth"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	"math/rand"
```
Please put these back at the top of the import block. See https://github.com/golang/go/wiki/CodeReviewComments#imports

roberthbailey commented:
@jayunit100 - Is this the first step towards being able to run other examples through e2e tests?

jayunit100 commented:
@roberthbailey yup ! thats the purpose of it.

jayunit100 commented:
did a quick push and cleaned up.  
ill take another look in the morning if i missed anything.

jayunit100 commented:
need to run gofmt, will do that shortly

jayunit100 commented:
yay build passed ; .... any thoughts on the change; or shall i rebase this? i separated the commits into separate to ease the review ; or keep as-is.

after, ill start integrating e2es from `examples`

roberthbailey commented:
Sorry for the delay @jayunit100. I did have a couple more comments. I'll try to get them to you tonight or tomorrow morning.

roberthbailey commented:
@jayunit100 - This isn't your fault, since the e2e tests were already all in one package, but I dislike how sprawlingly large this package is getting. Having everything in the same package means that we can't count on the compiler to help enforce any encapsulation between different test modules. I don't think we need to fix it in this PR, but as you start working on the examples e2e tests I'd like to see us move to multiple Go packages for different segments of the e2e tests.

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+	}
+}
+
+type validatorFn func(c *client.Client, podId string) (interface{}, error)
```
You never use the first return value from the validator function, so why have two return values?

roberthbailey replied:
s/podId/podID

roberthbailey replied:
I don't see why we'd need to pass the client back to the validatorFn. This can probably be simplified to `type validatorFn func(podID string) error`.

jayunit100 replied:
so --- the theoretical input to a validator is (1) a client, whose state we are validating and (2) a pod Id.  in some cases (2) is not required, but i can't imagine a case where (1) is never required ?  I'm trying to stick to the "pure" implementation where the validator does *nothing* but validate the cluster state, i.e., it doesnt create a client or do any other work. does that make sense.

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+
+type validatorFn func(c *client.Client, podId string) (interface{}, error)
+
+/**
```
Comments for a function should start with the function name. Also, it's more typical to use line comments rather than block comments for documenting functions. 

See http://golang.org/doc/effective_go.html#commentary

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+	// This template is unit-tested in kubectl, so if you change it, update the unit test.
+	//
+	// You can read about the syntax here: http://golang.org/pkg/text/template/
+	containername = "update-demo"
```
Remove this line.

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+
+			//call the generic validator function here
+			// this might validate for example, that (1) getting a url works and (2) url is serving correct content.
+			_, err := validator(c, podID)
```
The if check should be inlined here, e.g. `if _, err := validator(c, podID); err != nil { ... }`

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+			Logf("%s is verified up and running", podID)
+			runningPods = append(runningPods, podID)
+		}
+		//If we reach here, then all our checks passed
```
Comments should be full sentences (https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences) and you should leave a space between the // and the beginning of the comment.

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+			return
+		}
+	}
+	//Reaching here means that one of more checks failed multiple times.  Assuming its not a race condition, something is broke.
```
s/broke/broken

roberthbailey commented on test/e2e/kubectl.go:
```diff
@@ -17,17 +17,13 @@ limitations under the License.
-	"fmt"
-	"os/exec"
+	"errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
```
Move this import into it's own import group below the standard imports and above the ginkgo import (https://github.com/golang/go/wiki/CodeReviewComments#imports).

roberthbailey commented on test/e2e/kubectl.go:
```diff
@@ -153,82 +152,14 @@ func makeRequestToGuestbook(c *client.Client, cmd, value string) (string, error)
 }
 
-func getData(c *client.Client, podID string) (*updateDemoData, error) {
+/**
```
Please change this into line comments and begin the comment with the name of the function (http://golang.org/doc/effective_go.html#commentary).

roberthbailey commented on test/e2e/kubectl.go:
```diff
@@ -153,82 +152,14 @@ func makeRequestToGuestbook(c *client.Client, cmd, value string) (string, error)
+/**
+*  Get data for the update demo
+ */
+func getUDData(c *client.Client, podID string) (interface{}, error) {
```
If validateFn no longer takes a client as the first parameter, you'll probably want to change this into a function that takes a client and returns a validateFn:
```
func validateUpdateDemoPod(c *client.Client) validateFn {
  return func(podID string) error {
     ...
  }
}
```

roberthbailey commented on test/e2e/kubectl.go:
```diff
@@ -111,6 +107,9 @@ var _ = Describe("kubectl", func() {
 })
 
+/**
+* TODO this should leverage test utils file, rather than using its own waitForGuestbookResp implementation.
```
I don't understand this TODO. The wait here is different because it's waiting for an application specific response rather than a pod to be running or validating that a controller is in a specific state.

jayunit100 commented:
Hi okay, just saw these. let me look now.

jayunit100 commented:
pushed updates.  
- I fixed all the comments and
-  inlined the if/else stuff however, 
- I kept the **client** functionality in the `validatorFn`.  
thanks again 4 the feeback !  let me know if any  suggestions on a better validatorFn

jayunit100 commented:
okay, looks like its passing now.  im happy to implement more changes if anything comes to mind, as this will be the basis for a variety of important new tests! otherwise, let me know and ill rebase :)

roberthbailey commented on test/e2e/util.go:
```diff
@@ -182,3 +186,117 @@ func randomSuffix() string {
+	// helpful.
+	// This template is unit-tested in kubectl, so if you change it, update the unit test.
+	// You can read about the syntax here: http://golang.org/pkg/text/template/.
+	containername = "update-demo"
```
Still need to delete this line, otherwise the containername arg isn't doing anything.

roberthbailey commented:
Looking good. Just one last comment and then can you please squash your commits before I merge? Thanks.

jayunit100 commented:
okay ! I just did a final cleanup and push.

roberthbailey commented:
Thanks! LGTM. Will merge on green.

jayunit100 commented:
okay, repushed the rebase, added a semicolon,  ran things back through `gofmt` etc.  looks like its okay now.

jayunit100 commented:
ok..   green again now :) ... i pushed a second time just to reboot the builder....  same code from yesterday, btw, just rebased (and added one semicolon on line 172 of kubectl.sh)

roberthbailey commented:
@jayunit100 - Sorry, I had to revert your change because it caused our e2e tests to start failing consistently. Can you try again (I'll manually run the e2es before merging if you aren't able to do so).

jayunit100 commented:
no problem ! ill reissue it shortly....   I now have a good test
environment setup for running the e2es,so i'll make sure its functional
before PR'ing again.

On Thu, Mar 26, 2015 at 3:35 PM, Robert Bailey <notifications@github.com>
wrote:

> @jayunit100 <https://github.com/jayunit100> - Sorry, I had to revert your
> change because it caused our e2e tests to start failing consistently. Can
> you try again (I'll manually run the e2es before merging if you aren't able
> to do so).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5773#issuecomment-86684943>
> .
>



-- 
Jay Vyas
http://jayunit100.blogspot.com

roberthbailey commented:
Thanks. Please assign it to me once you send it out for review.

jayunit100 commented:
ahhh i see unmarshalling json into a byte array fails if vars are upcase.
i guess all go vars are sposed to be uppercase when using reflection.  I
didn't realize this.  ive got the issue worked out and will be pasting test
outputs with a new pr shortly, once all the tests indeed pass.

On Thu, Mar 26, 2015 at 6:34 PM, Robert Bailey <notifications@github.com>
wrote:

> Thanks. Please assign it to me once you send it out for review.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5773#issuecomment-86741544>
> .
>



-- 
Jay Vyas
http://jayunit100.blogspot.com

jayunit100 commented:
s/are/aren't.... :)

On Thu, Mar 26, 2015 at 6:51 PM, Jay Vyas <jayunit100@gmail.com> wrote:

> ahhh i see unmarshalling json into a byte array fails if vars are upcase.
> i guess all go vars are sposed to be uppercase when using reflection.  I
> didn't realize this.  ive got the issue worked out and will be pasting test
> outputs with a new pr shortly, once all the tests indeed pass.
>
> On Thu, Mar 26, 2015 at 6:34 PM, Robert Bailey <notifications@github.com>
> wrote:
>
>> Thanks. Please assign it to me once you send it out for review.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5773#issuecomment-86741544>
>> .
>>
>
>
>
> --
> Jay Vyas
> http://jayunit100.blogspot.com
>



-- 
Jay Vyas
http://jayunit100.blogspot.com

roberthbailey commented:
The vars must be uppercase for them to be settable from outside the current package, and the json unmarshaling code is running in a different package. One interesting side effect is that you can make the struct itself non-exported (starting with a lowercase letter) with exported variables and the marshaller can still set the variables via reflection.
